### ISSUE TITLE
Segfault when trying to connect (Linux)

### ISSUE THREAD

**Issue by dciotoli** [TimelordUK/node-sqlserver-v8#168]
OS: CentOS 7.6
Compiler: g++ 8.3.1
Node: version 12.18.3

Code executed:
```
const sql = require('msnodesqlv8');
const connStr = 'DRIVER={ODBC Driver 17 for SQL Server};SERVER=<my server>;DATABASE=<my database>;Trusted_Connection=yes';
const query = 'SELECT 1'
sql.query(connStr, query, (err, res) => { assert.ifError(err); console.log(res); });
```

**Comment by TimelordUK** [comment]
I have not tries trusted sign on via Linux - what happens if you choose say a non existant user/password - any error back would mean this module is at least talking to the microsoft ODBC driver - which version of that have you installed?

I am assuming you have compiled the code locally?

for example does it crash with something like below - if it does I would assume we are not even getting past initialisation with the ODBC driver which implies an incompatibility/installation issue.

const connectionString = 'Driver={SQL Server Native Client 17.0}; database=node; Server=<server>; UID=linux; PWD=linux'

one thing you could try is rebuild in debug mode

node-gyp -d rebuild

change lib/connection to use the Debug you just built

  const cppDriver = require('../build/Release/sqlserverv8.node')

launch debugger with node

gdb node

then in debugger run the tester

run ./test.js

*boom* blows up

get a full stack trace

bt

- post the full stack here to see where it is failing.

**Comment by dciotoli** [comment]
I've got ODBC 17 installed -- able to connect via trusted sign in w/ pyodbc in python. Will give compilation with debugging flags a try today and see what stack trace I get when things go boom

**Comment by dciotoli** [comment]
Might have something to do with the ODBC installation, got this stacktrack:

![Screen Shot 2020-08-04 at 5 08 20 PM](https://user-images.githubusercontent.com/6808302/89345290-27794c80-d675-11ea-95fe-60991e0f31bc.png)

**Comment by TimelordUK** [comment]
well obviosuly its going down in SSL

weirdly looking at my own distribution i dont see a dependency in ldd on libmsodbcsql-17.so - I include dependencies i see from Ubuntu.  I do note i seem to have ssl3 installed. Is it possible to see exactly what ssl you load when running ODBC Python with this driver

:/usr/lib$ find . -name "*ssl*.so" -print
./python3/dist-packages/cryptography/hazmat/bindings/_openssl.abi3.so
./x86_64-linux-gnu/libssl3.so
./python3.6/lib-dynload/_ssl.cpython-36m-x86_64-linux-gnu.so
./python2.7/lib-dynload/_ssl.x86_64-linux-gnu.so
./firefox/libssl3.so

/usr/lib$ ldd /usr/lib/libmsodbcsql-17.so
	linux-vdso.so.1 (0x00007ffd843b6000)
	libdl.so.2 => /lib/x86_64-linux-gnu/libdl.so.2 (0x00007fc1b0f6b000)
	librt.so.1 => /lib/x86_64-linux-gnu/librt.so.1 (0x00007fc1b0d63000)
	libodbcinst.so.2 => /usr/lib/x86_64-linux-gnu/libodbcinst.so.2 (0x00007fc1b0b48000)
	libkrb5.so.3 => /usr/lib/x86_64-linux-gnu/libkrb5.so.3 (0x00007fc1b0872000)
	libgssapi_krb5.so.2 => /usr/lib/x86_64-linux-gnu/libgssapi_krb5.so.2 (0x00007fc1b0627000)
	libstdc++.so.6 => /usr/lib/x86_64-linux-gnu/libstdc++.so.6 (0x00007fc1b029e000)
	libm.so.6 => /lib/x86_64-linux-gnu/libm.so.6 (0x00007fc1aff00000)
	libgcc_s.so.1 => /lib/x86_64-linux-gnu/libgcc_s.so.1 (0x00007fc1afce8000)
	libpthread.so.0 => /lib/x86_64-linux-gnu/libpthread.so.0 (0x00007fc1afac9000)
	libc.so.6 => /lib/x86_64-linux-gnu/libc.so.6 (0x00007fc1af6d8000)
	/lib64/ld-linux-x86-64.so.2 (0x00007fc1b1586000)
	libk5crypto.so.3 => /usr/lib/x86_64-linux-gnu/libk5crypto.so.3 (0x00007fc1af4a6000)
	libcom_err.so.2 => /lib/x86_64-linux-gnu/libcom_err.so.2 (0x00007fc1af2a2000)
	libkrb5support.so.0 => /usr/lib/x86_64-linux-gnu/libkrb5support.so.0 (0x00007fc1af097000)
	libkeyutils.so.1 => /lib/x86_64-linux-gnu/libkeyutils.so.1 (0x00007fc1aee93000)
	libresolv.so.2 => /lib/x86_64-linux-gnu/libresolv.so.2 (0x00007fc1aec78000)

**Comment by TimelordUK** [comment]
actually looking at what is actually loaded into my node process using the library is as follows

:~$ lsof -p 8163 | grep -i "/lib"
node    8163 stephen  mem       REG    8,1  2917216 131141 /usr/lib/x86_64-linux-gnu/libcrypto.so.1.1
node    8163 stephen  mem       REG    8,1   577312 131228 /usr/lib/x86_64-linux-gnu/libssl.so.1.1
node    8163 stephen  mem       REG    8,1    47568 679367 /lib/x86_64-linux-gnu/libnss_files-2.27.so
node    8163 stephen  mem       REG    8,1    97176 679364 /lib/x86_64-linux-gnu/libnsl-2.27.so
node    8163 stephen  mem       REG    8,1    47576 679369 /lib/x86_64-linux-gnu/libnss_nis-2.27.so
node    8163 stephen  mem       REG    8,1    39744 679365 /lib/x86_64-linux-gnu/libnss_compat-2.27.so
node    8163 stephen  mem       REG    8,1    14432 282750 /usr/lib/x86_64-linux-gnu/gconv/UTF-16.so
node    8163 stephen  mem       REG    8,1   101168 679373 /lib/x86_64-linux-gnu/libresolv-2.27.so
node    8163 stephen  mem       REG    8,1    14256 662286 /lib/x86_64-linux-gnu/libkeyutils.so.1.5
node    8163 stephen  mem       REG    8,1    43616 142831 /usr/lib/x86_64-linux-gnu/libkrb5support.so.0.1
node    8163 stephen  mem       REG    8,1    14248 656812 /lib/x86_64-linux-gnu/libcom_err.so.2.1
node    8163 stephen  mem       REG    8,1   199104 142823 /usr/lib/x86_64-linux-gnu/libk5crypto.so.3.1
node    8163 stephen  mem       REG    8,1   305456 142658 /usr/lib/x86_64-linux-gnu/libgssapi_krb5.so.2.2
node    8163 stephen  mem       REG    8,1   877056 142829 /usr/lib/x86_64-linux-gnu/libkrb5.so.3.3
node    8163 stephen  mem       REG    8,1   608957 132616 /usr/lib/x86_64-linux-gnu/libodbcinst.so.2.0.0
node    8163 stephen  mem       REG    8,1    31680 679374 /lib/x86_64-linux-gnu/librt-2.27.so
node    8163 stephen  mem       REG    8,1  2108784 133306 /opt/microsoft/msodbcsql17/lib64/libmsodbcsql-17.5.so.2.1
node    8163 stephen  mem       REG    8,1  2030544 679357 /lib/x86_64-linux-gnu/libc-2.27.so
node    8163 stephen  mem       REG    8,1   144976 679372 /lib/x86_64-linux-gnu/libpthread-2.27.so
node    8163 stephen  mem       REG    8,1    96616 656475 /lib/x86_64-linux-gnu/libgcc_s.so.1
node    8163 stephen  mem       REG    8,1  1700792 679361 /lib/x86_64-linux-gnu/libm-2.27.so
node    8163 stephen  mem       REG    8,1  1594864 132396 /usr/lib/x86_64-linux-gnu/libstdc++.so.6.0.25
node    8163 stephen  mem       REG    8,1    14560 679360 /lib/x86_64-linux-gnu/libdl-2.27.so
node    8163 stephen  mem       REG    8,1   170960 679353 /lib/x86_64-linux-gnu/ld-2.27.so
node    8163 stephen  mem       REG    8,1    26376 282755 /usr/lib/x86_64-linux-gnu/gconv/gconv-modules.cache

**Comment by dciotoli** [comment]
Results of the lsof command to see what gets loaded into both the node and python processes:

Node
<img width="651" alt="Screen Shot 2020-08-06 at 11 46 48 AM" src="https://user-images.githubusercontent.com/6808302/89552978-e69b4800-d7da-11ea-88b8-a281ad2cff99.png">

Python
![Screen Shot 2020-08-06 at 12 00 58 PM](https://user-images.githubusercontent.com/6808302/89554334-b8b70300-d7dc-11ea-915a-ebb278b0f574.png)

It seems like they're both loading the same libssl library (version 1.0.2k)

**Comment by TimelordUK** [comment]
looking at the pyodbc code

https://github.com/mkleehammer/pyodbc/blob/master/src/connection.cpp

does not look like they do anything fancy on connection.  We could try (if you have any spare time) playing around with OdbcConnection.cpp in theory we should be able to get it more or less calling same.  We will be calling the wide char version as i include the unicode define - i.e. the W variant of the open call.

For exampe what happens if you comment out in OdbcConnection.cpp the InitializeEnvironment one SQL_CP_ONE_PER_HENV line.

Both versions seem pretty similar in the open call - i seem to remember from stack trace the connection string looked valid.

**Comment by dciotoli** [comment]
Tried commenting it out, no luck. Still segfaulting at the same place in libssl.

**Comment by TimelordUK** [comment]
Thanks, looking in debugger under Linux I do see a problem in this area - the character vector is not sized correctly, looks like I had the wrong defines for LINUX set. Unless you can checkout the latest trunk, perhaps best waiting for a new build - I cannot guarantee this will fix it but with settings as they are we are handing a string incorrectly formatted, seems like I get away with it on Ubuntu.

**Comment by TimelordUK** [comment]
please give 2.0.2 a try when you get time.

**Comment by dciotoli** [comment]
Just tried out 2.0.2, still no luck.

**Comment by TimelordUK** [comment]
Hi,

I just built a Debian 10 VM and installed ms sql ODBC latest which looks to be 17.6 - i then did nothing other than install msnodesqlv8, compile it and it seems to run OK

I am not sure what else I can do to help



```cmd
@debian:~/dev/sql/node_modules/msnodesqlv8$ node runtest.js -u -t query
{ _: [], u: true, t: 'query' }
use conn_str Driver={ODBC Driver 17 for SQL Server}; database=node; Server=; UID=linux; PWD=linux


  query
    ✓ test retrieving a string with null embedded (425ms)
    ✓ test retrieving a large decimal as a string (83ms)
    ✓ multiple results from query in callback
    ✓ verify empty results retrieved properly
    ✓ object_name query  (1508ms)
    ✓ select nulls union all nulls
    ✓ test function parameter validation (93ms)
    ✓ test retrieving a LOB string larger than max string size
    ✓ query with errors
    ✓ simple query
    ✓ simple rawFormat query
    ✓ simple query of types like var% (210ms)
    ✓ streaming test
    ✓ serialized queries
    ✓ query with errors
    ✓ multiple results from query in events
    ✓ boolean return value from query
    ✓ test retrieving a non-LOB string of max size
    ✓ test retrieving an empty string
    ✓ test function parameter validation (131ms)
    ✓ verify metadata is retrieved for udt/geography types (1024ms)


  21 passing (4s)

0
@debian:~/dev/sql/node_modules/msnodesqlv8$ cat /etc/debian_version
10.5
@debian:~/dev/sql/node_modules/msnodesqlv8$ ls -l /lib/*17*.so
lrwxrwxrwx 1 root root 57 Aug  7 20:35 /lib/libmsodbcsql-17.so -> /opt/microsoft/msodbcsql17/lib64/libmsodbcsql-17.6.so.1.1

```

**Comment by dciotoli** [comment]
All good, thanks for all the effort you've been putting in. I'm going to run a similar experiment to see if I can replicate the issue in a virtual machine -- will update here with results.

**Comment by dciotoli** [comment]
solved by upgrading msodbcsql17 to the latest version (was getting segfault on 17.3, fixed on 17.6)
